bsyncviewer/lib/validator/workflow.py

```diff
@@ -112,62 +112,209 @@ def validate_all(self):
+            error = None
+            req_str = ", "
+            req_str = req_str.join(req_paths)
+            if state == 3:
+                # Required-1
+                if match_cnt > 1:
+                    msg = 'Required-1 error: multiple elements found, but there should be only one of the following: ' + req_str
+                    error = {'path': group_level, 'message': msg}
+                elif match_cnt == 0:
+                    msg = 'Required-1 error: one of the following elements is required: ' + req_str
+                    error = {'path': group_level, 'message': msg}
+            elif state == 4 and match_cnt > 1:
+                # Optional-1, but more than 1 selected
+                msg = 'Optional-1 error: multiple elements found, but there should be 0 or 1 of the following: ' + req_str
+                error = {'path': group_level, 'message': msg}
+            return error
+
+        for attr in attrs:
+            if group_num != attr.group_id and group_num > 0:
+                error = _error_check(state, match_cnt, group_level, req_paths)
+                if error is not None:
+                    errors.append(error)
+
+            # reset group number and match count
+            if group_num != attr.group_id:
+                group_num = attr.group_id
+                match_cnt = 0
+                state = attr.state
+                group_level = attr.grouping_level
+                req_paths = []
+
+            # find a match
+            req_paths.append(attr.attribute.path)
+            found, the_path = self.find_path_in_schema(attr.attribute.path)
+            if found:
+                match_cnt += 1
+
+        # error_check for last group
+        error = _error_check(state, match_cnt, group_level, req_paths)
+        if error is not None:
+            errors.append(error)
+
+        return errors
+
     def process_use_case(self, use_case):
         results = OrderedDict()
         results['valid'] = True
         results['errors'] = []
 
         # get all attributes of use case where state > 0
-        # 1 = Optional, 2 = Required
+        # 1 = Optional, 2 = Required, 3 = Required-1, 4 = Optional-1.
+        # validate allowable values and required paired elements in this loop, but not groups
         attrs = UseCaseAttribute.objects.filter(state__gt=0, use_case=use_case)
-        # print("NUMBER OF ATTRIBUTES RETRIEVED: {}".format(attrs.count()))
 
         # traverse use case attributes (not xml)
         for attr in attrs:
             # try to find it
             paths = copy.deepcopy(attr.attribute.path.split("."))
-            loopingVar = copy.deepcopy(self.xml)
+
             # special case: UDF
+            # print("ATTR PATH: {}".format(attr.attribute.path))
             if 'UserDefinedField.FieldName' in attr.attribute.path:
                 # retrieve all FieldNames from UseCaseUDF
                 udfFields = UseCaseUDF.objects.filter(use_case_attribute=attr)
```

Comment:
your JS syntax is creeping into python :D... udf_fields.

Reply:
m_whatsWrongWith_this_Syntax

Reply:
I don't see a problem.